### PR TITLE
[BLOG]  Integrating Hypothesis into Sympy

### PR DESCRIPTION
**Pull request by dianetc**
## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] Blog post featured image is in PNG or JPEG format, **not** SVG.
- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [x] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

**Comment by dianetc** [comment]
I'm assuming the CI is failing partly because 
```
featuredImage:
  src: 
  alt: 
hero:
  imageSrc: 
  imageAlt: 
```
in my meta section is empty, but I have no figures/images to add.

**Comment by asmeurer** [comment]
This post looks good @dianetc. It looks like the preview build failed, but I'm not sure how to fix it. I think there's some issue with the metadata markup. 

One suggestion I would make is that the Circle example feels a little contrived. All the test does is test for the exact thing that's in the code, which doesn't really demonstrate why you'd want to use a property test.

**Comment by kgryte** [comment]
Would it make sense to show before and after snippets of SymPy tests?

Personally, I'd like to see a bit more discussion in this post regarding Hypothesis integration in SymPy, rather than requiring me to consult external resources. E.g., I could imagine showing SymPy tests before Hypothesis, maybe even including an image of passing tests, and then showing SymPy tests post Hypothesis with various failures identified.

I'd also be interested in knowing how one goes about choosing which strategies to leverage and whether SymPy required any custom strategies.

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+---
+title: 'Integrating Hypothesis into Sympy'
```
Comment (kgryte) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:2]: nit: s/Sympy/SymPy/ here and elsewhere.

**Comment by lucascolley** [comment]
> It looks like the preview build failed, but I'm not sure how to fix it.

This should go away once us interns are added to Storyblok and author slugs are generated. Pavithra is aware :)

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  
+* Hypothesis is **_not_** just a bug finder, it helps protect against _future_ bugs. Hypothesis disallows the existence of latent bugs which increases trust in the current implementation of whatever function is being tested. Hypothesis may also reveal weird design patterns. 
+
+# Integrating Hypothesis into SymPy
+
+SymPy is an ideal library for property based testing so integration was painless.
+
+### What Has Changed in SymPy?
+
+Hypothesis is now a required _testing_ dependency of SymPy.  Property based tests can be created in the _test_hypothesis.py_ file in respective tests directories (More on this [here](https://github.com/sympy/sympy/blob/master/doc/src/contributing/new-contributors-guide/writing-tests.md#hypothesis-testing)). An example testing file using hypothesis can be found in  [ntheory/tests](https://github.com/sympy/sympy/blob/master/sympy/ntheory/tests/test_hypothesis.py)
```
Comment (pavithraes) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:97]:
```suggestion
Hypothesis is now a required _testing_ dependency of SymPy. Property based tests can be created in the `test_hypothesis.py` file in respective tests directories (more details in the [new contributor documentation](https://github.com/sympy/sympy/blob/master/doc/src/contributing/new-contributors-guide/writing-tests.md#hypothesis-testing)). An example testing file using hypothesis can be found in  [ntheory/tests](https://github.com/sympy/sympy/blob/master/sympy/ntheory/tests/test_hypothesis.py).
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+### Utilizing Hypothesis in SymPy
+
+Hypothesis was able to find various bugs and code design flaws. Below I will highlight two:
+* The resultant function returning [incorrect answers](https://github.com/sympy/sympy/issues/25406). While this bug ended up not needing to be resolved, it did  reveal the utility in having Hypothesis check [consistency between implementations](https://github.com/sympy/sympy/issues/25406#issuecomment-1652243538) of the same function.
```
Comment (pavithraes) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:101]:
```suggestion
* The [`resultant` function returning incorrect answers](https://github.com/sympy/sympy/issues/25406). While this bug ended up not needing to be resolved, it did  reveal the utility in having Hypothesis check [consistency between implementations](https://github.com/sympy/sympy/issues/25406#issuecomment-1652243538) of the same function.
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  
+* There were various issues with the lowest common multiple function ([here](https://github.com/sympy/sympy/issues/25624), [here](https://github.com/sympy/sympy/pull/25636), and [here](https://github.com/sympy/sympy/pull/25517#issuecomment-1714474991)), the biggest being when the lcm should make use of the integer implementation vs polynomial implementation (when the defined polynomial is essentially an integer).
```
Comment (pavithraes) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:103]:
```suggestion
* There were various issues with the lowest common multiple (LCM) function (notes in [issue#25624](https://github.com/sympy/sympy/issues/25624), [PR#25636](https://github.com/sympy/sympy/pull/25636), and [PR#25517](https://github.com/sympy/sympy/pull/25517#issuecomment-1714474991)), the biggest being when the LCM should make use of the integer implementation vs polynomial implementation (when the defined polynomial is essentially an integer).
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+title: 'Integrating Hypothesis into SymPy'
+authors: [diane-tc]
+published: September 20, 2023
+description: 'Gives an introduction to the utility of hypothesis in SymPy'
+category: [Property Based Testing]
+featuredImage:
+  src: /posts/integrating-hypothesis-into-sympy/blog_feature_org.png
+  alt: 'An illustration of a brown hand holding up a microphone, with some graphical elements highlighting the top of the microphone.'
+hero:
+  imageSrc: /posts/integrating-hypothesis-into-sympy/blog_hero_var2.svg
```
Comment (pavithraes) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:11]: Note - I don't see the image on my local website build, although everything seems right here. We need to verify if this works in the preview deployment.

**Review by pavithraes — CHANGES_REQUESTED** [review]
Thanks for the post, @dianetc! I've shared some notes and suggestions. :)

I've also added you to Storyblok, so your next commit should hopefully trigger preview deployment without errors.

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  imageAlt: 'An illustration of a brown hand holding up a microphone, with some graphical elements highlighting the top of the microphone.' 
+---
+This summer I interned at Quansight Labs with a focus of intergrating [Hypothesis](https://github.com/HypothesisWorks/hypothesis/) into the testing suit of [SymPy](https://github.com/sympy/sympy). The primary [pull request to add Hypothesis to the test infrastructure](https://github.com/sympy/sympy/pull/25428) to complete this was simple. The primary challenges lied thereafter: questions around the _utility of hypothesis_ and its appropriate usage arose.
```
Comment (melissawm) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:14]:
```suggestion
This summer I interned at Quansight Labs with a focus of integrating [Hypothesis](https://github.com/HypothesisWorks/hypothesis/) into the testing suit of [SymPy](https://github.com/sympy/sympy). The primary [pull request to add Hypothesis to the test infrastructure](https://github.com/sympy/sympy/pull/25428) to complete this was simple. The primary challenges lied thereafter: questions around the _utility of hypothesis_ and its appropriate usage arose. 
```
It also may make sense to remove one of the `primary` instances just to avoid repetition.

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+* Experience Integrating into SymPy
+
+If you wish to follow the examples in the blog post, you will need to have the latest version of sympy (on master) and you need to install Hypothesis via:
+
+```bash
+pip install hypothesis
+```
+# What is Property Based Testing?
+
+Property based testing (PBT) is a technique where instead of testing individual test cases, you specify properties you which to hold true for a range of inputs. These properties are then tested against automatically generated test data. PBT uses logical properties over generated test data to facilitate broad, robust testing that can expose edge cases not easily found via traditional test cases.
```
Comment (melissawm) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:29]:
```suggestion
Property based testing (PBT) is a technique where instead of testing individual test cases, you specify properties you wish to hold true for a range of inputs. These properties are then tested against automatically generated test data. PBT uses logical properties over generated test data to facilitate broad, robust testing that can expose edge cases not easily found via traditional test cases.
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+```python
+from sympy.abc import x
+def test_degree():
+    f = Poly(x**2 + 1, x)
+    g = Poly(x**3 + 1, x)
+    h = f*g
+    assert h.degree() == f.degree() + g.degree()
+```
+Notice, we are limited in how many various combinations of _f_ and _g_ we can test. It would be better if we could fix a property and have a library automatically generate interesting test cases and run them for us. We would no longer need to worry about thinking up input/output pairs. This would increase our **_trust_** in the implementation.
```
Comment (melissawm) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:48]:
```suggestion
Notice we are limited in how many various combinations of _f_ and _g_ we can test. It would be better if we could fix a property and have a library automatically generate interesting test cases and run them for us. We would no longer need to worry about thinking up input/output pairs. This would increase our **_trust_** in the implementation.
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+# What is Hypothesis?
+ > [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) is a Python library for creating unit tests which are simpler to write and more powerful when run, finding edge cases in your code you wouldn’t have thought to look for. It is stable, powerful and easy to add to any existing test suite.
+> Now, let's test out the property above using Hypothesis:
+  
+  ```python
```
Comment (melissawm) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:54]: I can't see the preview, but the indentation here may cause problems there. You probably want to have
```suggestion
```python
```

File: apps/labs/posts/integrating-hypothesis-into-sympy.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+from hypothesis import given, strategies as st
+
+@given(f = polys(), g = polys())
+def test_degree(f, g):
+    h = f * g
+    assert h.degree() == f.degree() + g.degree()
+```
+Note, that here `polys()` is custom-built for sympy and generates a random polynomial with integer coefficeints. It is not a built-in Hypothesis strategy.
```
Comment (melissawm) [inline apps/labs/posts/integrating-hypothesis-into-sympy.md:62]:
```suggestion
Note, that here `polys()` is custom-built for sympy and generates a random polynomial with integer coefficients. It is not a built-in Hypothesis strategy.
```

**Review by melissawm — APPROVED** [review]
Hi @dianetc - this was very informative and interesting! I left a few comments (mostly typos) but overall this looks great to me. Cheers!

**Comment by melissawm** [comment]
Oops! @dianetc I am assuming this is not what you wanted - could you restore the blog post? Feel free to open a new pull request if that is easier. Thanks!
